### PR TITLE
ipc: use highest coop priority for IPM workqueue

### PR DESCRIPTION
**Pull request by ankuns**
Signed-off-by: Andrzej Kuroś <andrzej.kuros@nordicsemi.no>

### CODE REVIEW

**Review by MarekPorwisz — APPROVED** [review]

File: subsys/ipc/rpmsg_service/rpmsg_backend.c [hunk 1]
```diff
@@ -43,7 +43,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_RPMSG_SERVICE_LOG_LEVEL);
 #define IPM_WORK_QUEUE_STACK_SIZE CONFIG_RPMSG_SERVICE_WORK_QUEUE_STACK_SIZE
 
 #if IS_ENABLED(CONFIG_COOP_ENABLED)
-#define IPM_WORK_QUEUE_PRIORITY -1
+#define IPM_WORK_QUEUE_PRIORITY K_PRIO_COOP(0)
```
Comment (carlocaione) [inline subsys/ipc/rpmsg_service/rpmsg_backend.c:46]: Just use `K_HIGHEST_APPLICATION_THREAD_PRIO` instead.

Comment (ankuns) [reply subsys/ipc/rpmsg_service/rpmsg_backend.c]: fixed

File: subsys/ipc/rpmsg_service/rpmsg_backend.c [hunk 2]
```diff
@@ -43,7 +43,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_RPMSG_SERVICE_LOG_LEVEL);
 #define IPM_WORK_QUEUE_STACK_SIZE CONFIG_RPMSG_SERVICE_WORK_QUEUE_STACK_SIZE
 
 #if IS_ENABLED(CONFIG_COOP_ENABLED)
-#define IPM_WORK_QUEUE_PRIORITY -1
+#define IPM_WORK_QUEUE_PRIORITY K_HIGHEST_APPLICATION_THREAD_PRIO
 #else
 #define IPM_WORK_QUEUE_PRIORITY 0
 #endif
```
Comment (carlocaione) [inline subsys/ipc/rpmsg_service/rpmsg_backend.c:49]: Uhm, actually using `K_HIGHEST_APPLICATION_THREAD_PRIO` you can remove the whole `if` block now

Comment (ankuns) [reply subsys/ipc/rpmsg_service/rpmsg_backend.c:49]: Yes, I took minimum-change approach, push in minutes.

**Review by carlocaione — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by hubertmis — COMMENTED** [review]
I'm not convinced if setting the highest possible, often "cooperative" priority to a thread calling callbacks is the correct solution. We don't have any control if the callback called by this thread is implemented in "cooperative" way. It might be performing heavy crypto operation or even busy waiting for a hardware event. Because of this, the highest priority rpmsg workqueue thread can starve all other threads in the system, disrupting its operations.

I understand this was merged as a workaround to a problem with other thread which was configured as a cooperative, but was not implemented to work in a cooperative way. Perhaps we should fix the problem with the other thread which is blocking rpmsg_service workqueue instead of moving problem around increasing priorities or all threads to higher values than the misbehaving thread.

With this change merged, other threads which potentially can have higher priority than rpmsg_service workqueue cannot be properly configured, because the workqueue occupies the highest possible priority.

@carlocaione, @carlescufi, @ankuns Thoughts?

**Comment by ankuns** [comment]
Highest possible often "cooperative" and also often meta-irq priotity. Intentional. That's the point of the workaround.

As a rule of thumb the higher priority the shorter the handler should be. With merged PR it is possible to check in callback what is going to be executed. If a time-consuming operation then it can be passed to some lower priority. However the opposite is not possible when callback is executed on low priority. The callback simply never gets executed in time to even take such decision.

I agree that it would be better to fix in "the other thread" (or threads). For now we need this workaround.

@hubertmis  The `IPM_WORK_QUEUE_PRIORITY` could be Kconfig-urable. Would you be more convinced? Then it could be set to highest possible priority for the use case this fix is for.

**Comment by hubertmis** [comment]
> If a time-consuming operation then it can be passed to some lower priority

The goal of the workqueue inside rpmsg_service implementation is to pass execution context to some lower priority. I don't think we should lower a little in rpmsg_service and lower a little more in the callback. Lowering priority should be handled in rpmsg_service only, or in the callback code only. My preference is rpsmg_service, because it would not need the user of this module to worry about changing execution context.

> I agree that it would be better to fix in "the other thread" (or threads). For now we need this workaround.

I understand, but I think we should immediately plan proper fix and removal of this workaround. Possibly with moving workqueue to a preemtive priority level. This workaround is affecting every rpmsg_service user. Not only the failing application with the "misbehaving thread".

**Comment by carlocaione** [comment]
Can we just add a new Kconfig for the priority?

**Comment by hubertmis** [comment]
> Can we just add a new Kconfig for the priority?

We can, but I understand it would be another workaround instead of a proper fix.

**Comment by carlocaione** [comment]
> We can, but I understand it would be another workaround instead of a proper fix.

As always I don't think we should babysit the developer :) We should give the upper layer the tool, not the policy, so I do not consider the Kconfig option a workaround. Using the Kconfig the user can select the priority he wants and the priority he thinks is the right one for the implementation of his actual endpoint callback.

**Comment by hubertmis** [comment]
I'm curious what default value we'll provide for this Kconfig. I think it should be preemptive because of the callout we can't control. Also are we going to override the default in samples? I hope it won't be necessary

**Comment by carlocaione** [comment]
I would leave the default as it was before this change

**Pull request merged**